src/Workspaces/Core/Portable/Workspace/Host/DocumentService/Extensions.cs

```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    internal static class Extensions
+    {
+        public static bool CanApplyChange(this TextDocument document)
+        {
+            return document?.State.CanApplyChange() ?? false;
+        }
+
+        public static bool CanApplyChange(this TextDocumentState document)
+        {
+            return document?.Services.GetService<IDocumentOperationService>().CanApplyChange ?? false;
+        }
+
+        public static bool SupportDiagnostics(this TextDocument document)
+        {
+            return document?.State.SupportDiagnostics() ?? false;
+        }
+
+        public static bool SupportDiagnostics(this TextDocumentState document)
+        {
+            return document?.Services.GetService<IDocumentOperationService>().SupportDiagnostics ?? false;
```

Comment:
what is an example of a document that doesn't support diagnostics?

Reply:
razor won't support diagnostics for closed files.

Reply:
this code seems to be weird:
1. when are we passed a 'null' document?
2. it looked like in this PR that document services could be null.  So wouldn't .SupportsDiagnostics be something that could then throw?